handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -574,11 +575,17 @@ public SSLEngine engine() {
      */
     public String applicationProtocol() {
         SSLSession sess = engine().getSession();
-        if (!(sess instanceof ApplicationProtocolAccessor)) {
-            return null;
+        if (sess instanceof ApplicationProtocolAccessor) {
+            return ((ApplicationProtocolAccessor) sess).getApplicationProtocol();
+        }
+        if (Java9SslUtils.supportsAlpn(engine())) {
```

Comment:
I think you're referring to the problem where a class is statically linked against a class that is not present at runtime, such as in the case where javac is configured to link against the JDK8 standard library while producing bytecode compatible with Java 6. The advantage of using reflection in Java9SslUtils is that it is Java 6 compatible on all three levels (source, target, and library). In recent revisions of the PR, the static initializer also tolerates runtimes older than JDK9, so it is always safe to call supportsAlpn.